source/common/buffer/buffer_impl.cc

```diff
@@ -12,26 +12,39 @@
 namespace Envoy {
 namespace Buffer {
 
-// RawSlice is the same structure as evbuffer_iovec. This was put into place to avoid leaking
-// libevent into most code since we will likely replace evbuffer with our own implementation at
-// some point. However, we can avoid a bunch of copies since the structure is the same.
-static_assert(sizeof(RawSlice) == sizeof(evbuffer_iovec), "RawSlice != evbuffer_iovec");
-static_assert(offsetof(RawSlice, mem_) == offsetof(evbuffer_iovec, iov_base),
-              "RawSlice != evbuffer_iovec");
-static_assert(offsetof(RawSlice, len_) == offsetof(evbuffer_iovec, iov_len),
-              "RawSlice != evbuffer_iovec");
+/**
+ * Copy a Reservation to a RawSlice. The public interface of RawSlice makes specific
+ * fields immutable (the comments in the declaration of RawSlice explain why), so this
+ * is a convenience function that encapsulates the required casting.
+ * @param lhs the slice to be overwritten with a copy of rhs.
+ * @param rhs the reservation to copy.
+ */
+static inline void Copy(RawSlice& lhs, const Slice::Reservation& rhs) {
+  *(const_cast<void**>(&(lhs.mem_))) = rhs.mem_;
+  lhs.len_ = rhs.len_;
+}
 
-void OwnedImpl::add(const void* data, uint64_t size) { evbuffer_add(buffer_.get(), data, size); }
+void OwnedImpl::add(const void* data, uint64_t size) {
```

Comment:
Yes for mostly historical reasons, this exists before we have string_view in the codebase. prepend is introduced later so we only have string_view version.

Reply:
The problem I see with string_view is that it uses char* as its pointer type. The signedness of char is unspecified, so any client code that operates on binary protocols would have to reinterpret_cast between char* and uint8_t*. We'd end up with fewer static_casts inside the buffer impl, but at the cost of a lot of new reinterpret_casts throughout the filter and codec code.

Reply:
It might be useful to talk through the three cases of casts individually:

`reinterpret_cast`:
I just double-checked buffer_impl.cc, and all the reinterpret_casts are in the old implementation.

`static_cast`:
Most of the static_casts are there to turn void* provided by callers into uint8_t*. The equivalent code in the previous implementation uses C-style casts in the same spots: [libevent/buffer.c](https://github.com/libevent/libevent/blob/63c6669fd831088825dfc0000737a33853bb36e8/buffer.c). So my perspective has been that the static_casts make the new impl easier to read than the old impl.

The other uses of static_cast are in the move methods, to enforce the constraint that moves can only occur between the same buffer subclass. The old implementation does the same thing. I think your point about validating this with a runtime "ASSERT(dynamic_cast(..." is reasonable; I'll add that.

`const_cast`:
There are two const_casts in the new buffer implementation, and they're both there to support the change I made in `RawSlice`: replacing `void* mem_` with `void* const mem_`. That was a conscious tradeoff: making the field in the public struct `const` helps ensure that callers can't corrupt a reservation, since the reserve/commit model depends on the pointer not changing. The price of that external constness is a pair of const_casts inside the buffer impl. Alternatively, I could remove the `const` from the public `RawSlice` struct declaration and thus eliminate the need for the const_casts, but the tradeoff there would be less protection against callers corrupting the reservations. What do you think: is it cleaner to keep the const in the public interface (and incur the need for the const_casts internally), or to revert to the non-const public interface (and maybe just use a warning in the comments to persuade callers not to modify the `mem_` field)?

(If we could start over from scratch, the cleaner solution would be to replace `RawSlice` with a class containing private fields and public accessor methods. But doing that now would break people's third-party filter code.)

There used to be a lot more static_casts in the new implementation, but those went away with one of the prior revisions to this PR where I changed the internal slice API to use `uint8_t*` instead of `void*`.